pkg/reconciler/v1alpha1/channel/channel.go

```diff
@@ -0,0 +1,140 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// controllerAgentName is the string used by this controller to identify
+	// itself when creating events.
+	controllerAgentName       = "channel-default-controller"
+	channelReconciled         = "ChannelReconciled"
+	channelUpdateStatusFailed = "ChannelUpdateStatusFailed"
+)
+
+type reconciler struct {
+	client        client.Client
+	restConfig    *rest.Config
+	dynamicClient dynamic.Interface
+	recorder      record.EventRecorder
+}
+
+// Verify the struct implements reconcile.Reconciler
+var _ reconcile.Reconciler = &reconciler{}
+
+// ProvideController returns a Channel controller.
+// This Channel controller is a default controller for channels of all provisioner kinds
+func ProvideController(mgr manager.Manager) (controller.Controller, error) {
+	// Setup a new controller to Reconcile channel
+	c, err := controller.New(controllerAgentName, mgr, controller.Options{
+		Reconciler: &reconciler{
+			recorder: mgr.GetRecorder(controllerAgentName),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Watch channel events
+	// This controller is no-op when Channels are deleted
+	if err := c.Watch(
+		&source.Kind{Type: &v1alpha1.Channel{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.Funcs{
+			DeleteFunc: func(event.DeleteEvent) bool {
+				return false
+			},
+		}); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// This defaul channel reconciler will check if the channel is being watched by provisioner's channel controller
+// This will improve UX. See https://github.com/knative/eventing/issues/779
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	glog.Infof("Reconciling channel %v", request)
+	ch := &v1alpha1.Channel{}
+
+	// Controller-runtime client Get() always deep copies the object. Hence no need to again deep copy it
+	err := r.client.Get(context.TODO(), request.NamespacedName, ch)
+
+	if errors.IsNotFound(err) {
+		glog.Errorf("could not find channel %v\n", request)
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		glog.Errorf("could not fetch channel %v for %+v\n", err, request)
+		return reconcile.Result{}, err
+	}
+
+	err = r.reconcile(ch)
+
+	if err != nil {
+		glog.Warningf("Error reconciling channel: %v. Will retry.", err)
+		r.recorder.Eventf(ch, corev1.EventTypeWarning, channelUpdateStatusFailed, "Failed to update channel status: %v", request.NamespacedName)
+		return reconcile.Result{Requeue: true}, err
+	}
+	glog.Infof("Successfully reconciled channel %v", request.NamespacedName)
+	r.recorder.Eventf(ch, corev1.EventTypeNormal, channelReconciled, "Channel reconciled: %v", request.NamespacedName)
+	return reconcile.Result{Requeue: false}, nil
+}
+
+func (r *reconciler) reconcile(ch *v1alpha1.Channel) error {
+	ch.Status.InitializeConditions()
```

Comment:
And then here, we should not call InitializeConditions, just call `SetProvisionerNotInstalled` when you understand the answer.